Release 0.10.0 (#120)
kuzudb/kuzu-blog: src/content/post/2025-05-05-kuzu-v-0.10.0.md

```cypher
CALL neo4j_migrate(
    'host_name',
    'user_name',
    'password',
    ['node_label_1', 'node_label_2', ...],
    ['rel_label_1', 'rel_label_2', ...]
)
```
with the node and relationship labels you wanna migrate.

## Scan compressed CSV

Large CSV files are usually distributed in compressed formats. Kuzu now supports scanning directly from gzipped csv files without manual decompression.

Consider a `user.csv` file, its gzipped version `user.csv.gz` can be obtained by
```shell
gzip -k user.csv
```

The compressed file can then be read by Kuzu through
```cypher
LOAD FROM 'user.csv.gz' RETURN *;
```
It can also be used for bulk insertion
```cypher
COPY User FROM 'user.csv.gz';
```

## Android support

Kuzu now supports Android devices with ARMv8-A architecture through our Java API. Developers can now build mobile applications that leverage Kuzu's graph database capabilities directly on Android devices.
The precompiled binaries are compiled targetting [API level 21](https://developer.android.com/tools/releases/platforms).

## Performance Improvements

We present 

### Free Space Management

> Reusing disk space has been a nice but long missed feature in Kuzu.

Reviewer: ```suggestion
Reusing disk space has been a long missed feature in Kuzu.
```
The `nice` makes it sound a bit like we already have the feature, should probably either replace it with a more appropriate adjective or we can just remove it altogether

> It can prevent database size from bloating after frequent data modifications.

Reviewer: ```suggestion
It can prevent database size from becoming bloated after frequent data modifications.
```

We're now introducing the free space management mechanism that is able to reclaim free space in three key scenarios:
- When tables are dropped;
- When tuples within a node group are all deleted;
- When column chunks are rewritten due to updates or insertions.

> The space reclamation process occurs automatically during the CHECKPOINT phase, ensuring minimal impact on regular transactions.

Reviewer: ```suggestion
The space reclaimation process occurs automatically during the CHECKPOINT phase, ensuring minimal impact on regular transactions.
```

Note that:
- The current implementation focuses on managing free space in the `data.kz` file, which stores main data of node and rel tables. We plan to extend this functionality to index files in future releases.

> - We never shrink our `data.kz` file, but only keeps track of free space within the file and try to reuse in the future.

Reviewer: ```suggestion
- We never shrink our `data.kz` file. Instead, we keep track of free space within the file and try to reuse it during future allocations.
```

To demonstrate the effectiveness of our new free space management mechanism, we conducted a micro-benchmark using the LDBC dataset at scale factor 100.
We compared the disk space usage between v0.9.0 and v0.10.0 across three different workloads:
1. **Multi COPY**: We partition the `person` csv file into `x` partitions, each of which is saved as a csv file. We copy each partitioned csv file one at a time, and report the final size of `data.kz` file.
2. **Drop Tables**: We create a node table `person`, copy tuples into the table, drop the table, and perform `CHECKPOINT`. We repeat these operations for 5 times, and report the final size of `data.kz` file.
3. **Bulk Deletions**: We create a node table `person`, copy tuples into the table, delete all tuples from the table, and perform `CHECKPOINT`. We repeat these operations for 5 times, and report the final size of `data.kz` file.

| Workloads | 0.9.0 | 0.10.0 |
| --------- | ----- | ------ |

> | Multi COPY | 32MB | 20MB |

Reviewer: Numbers are for 10 partitions

## Scan compressed CSV

Our users commonly requested that their compressed CSV files be directly scanned or copied into Kuzu. We're happy to add support
for scanning or copying `.gzip.csv` files directly into Kuzu starting from this release.

Consider a `user.csv` file. Let's first zip this file to a new file named `user.csv.gz` as follows:

```shell
gzip -k user.csv
```

The compressed CSV file can then be scanned or copied into a Kuzu table:
```cypher
// scan
LOAD FROM 'user.csv.gz' RETURN *;
// copy
COPY User FROM 'user.csv.gz';
```

Storing compressed CSV files in Kuzu is useful when importing data from external systems that store
data in compressed CSV format, such as object stores. See [here](https://docs.kuzudb.com/import/csv/#compressed-csv-files)
for more details on this feature.

## Android support

We're excited to now support Android devices with ARMv8-A architecture through our Java API. Developers can
now build mobile applications that leverage Kuzu's graph database capabilities directly on Android devices.

> The precompiled binaries are compiled targetting [API level 21](https://developer.android.com/tools/releases/platforms).

Author: I rewrote this part a bit. It's mainly for people who want to develop Android applications with Kuzu.

whose size was linear to the number of nodes in relevant tables. This incurred a performance overhead when running a _small_ recursive query
on _very large_ database in which case, the allocation became the bottleneck. 

In v0.9.0, on LDBC100 with 220M `Comment` records, a simple index nested loop join query took about **1ms** but its equivalent recursive query takes **300ms** due to the large size of the `Comment` table.

```cypher
// Non-recursive
MATCH (:Comment {id:<id>})-[:replyOf]->(:Comment)
// Recursive
MATCH (:Comment {id:<id>})-[:replyOf*1..1]->(:Comment) // 300ms
``` 

In v0.10.0, we eliminate the bottleneck by starting with a small allocation and dynamically growing the data structure.
This brings the recursive query performance back to **1ms**, on par with the non-recursive query.

| Query | 0.9.0 | 0.10.0 |
| ----- | ----- | ------ |
| Non-recursive | 1ms   | 1ms    |
| Recursive     | 300ms | 1ms    |

That's a 300x improvement in performance!

### JSON scan

Importing JSON data into Kuzu is a common use case. In v0.10,0, we've significantly improved the performance

> of JSON scanning. Below, we show a micro benchmark that scans LDBC-100 dataset in JSON format.

Reviewer 2: Should be comment table

```cypher
INSTALL json;
LOAD json;  
// Scan json file
LOAD FROM 'ldbc100-file.json' RETURN *;

> ```

Reviewer 2: No this is not the correct query, i will modify the doc

|  | 1 thread | 2 threads | 4 threads |
|---------|----------|-----------|-----------|
| v0.9    | 190s    | 99s      | 52s      |
| v0.10   | 56s    | 27s      | 15s     |

> In general, you should see a significant performance improvement when scanning JSON files.

Reviewer 2: I think it only improves JSON scan/copy performance.
Neo4j migrate uses CSV right now because NEO4J apoc extension is much slower when exporting to json

All algorithms, except for that [(all) weighted shortest paths](http://docs.kuzudb.com/cypher/query-clauses/match#shortest-path) which is provided as built-in, are available via the [`algo` extension](http://docs.kuzudb.com/extensions/algos). 
Let's look at example usage of graph algorithms using the well-known LDBC SNB dataset.
The example beloew finds top 10 influencers based on the `(person)-[:knows]->(person)` subgraph and counts how many posts they have published.

```cypher
// Install graph algorthm extension
INSTALL algo;
LOAD algo;

// Create a projected graph with person and knows tables.
CALL project_graph(
    'KnowsGraph', 
    ['person'], 
    ['knows']
);

// Run page rank on projected graph to find the top 10 influencers.
// Then traverse `postHasCreator` to extract all posts
CALL page_rank('KnowsGraph')
WITH node AS person, rank
ORDER BY rank DESC
LIMIT 10
MATCH (person)<-[:postHasCreator]-(post:Post)
RETURN person, COUNT(*);
```

We first install and load the `algo` extension. Then, we create a projected graph from the `person` and `knows` tables,
on which the algorithm is executed. The projected graph is a subgraph that contains only the nodes and

> relationships specified. It is evaluated only when the algorithm is executed. Kuzu does not materialize projected

Reviewer 3: Do we want to include technical details of how projected graph is evaluated in the post? I would just point to doc.

Reviewer 4: I want to include it because from a reader's perspective, it's not immediately clear why a projected graph is necessary. Actually, we should drop a brief line what the purpose of the projected graph is. For an uninitiated user, they might expect that the graph algorithm just runs on the full graph, without this intermediate step. We shouldn't assume they know that.
